[Libbeat][New Processor] XML Decode

P1llus opened the pull request:
## What does this PR do?

This PR adds 2 components to `Libbeat`.
1. It adds a small XML Unmarshal helper function to the `common` folder in `Libbeat`, the reason this is added is so that not only processors can utilize it if/when needed, but also inputs. For example I plan to reuse this in `http_endpoint` to add XML support there as well. Other beats/inputs that would benefit from this is `Winlogbeat`.
The helper function expects a `[]byte` with valid XML object(s), and will return it as a struct reusing the names of the XML object(s).
It supports lists, arrays, nested fields, object identifiers (for example `<book seq="1">`).
If the XML is not valid it will return a proper error message describing why it failed.
For more information on supported formats, please see the included unit test files.

2. Second component is a `xmldecode` processor, **this is the first processor I made, so please review it carefully**. I have used `urldecode`, `mime_type` and the `decode_json_fields` processor as a guideline for this one.
For the supported objects and more information on supported formats, please see the included unit test files.

## Why is it important?

This adds the possibility to read XML files for file/log input, or decoding XML strings in existing message fields, for example XML strings in existing JSON fields.
Implements an easy helper function for other parts of beats to reuse if they want to Unmarshal XML into a struct.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #23366

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23678/detail/PR-23678/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23678/detail/PR-23678/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23678/detail/PR-23678/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23678/detail/PR-23678/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_23678.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #23678 opened`



* * Start Time: 2021-01-26T10:31:16.907+0000


* Duration: 13 min 25 sec


* Commit: c5c5870c258f15998e6a79a15f12f83ece9436e4




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->







### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-23678/detail/PR-23678/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Check for changes`
<ul>
<li>Took 0 min 3 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23678/runs/1/steps/174/log/?start=0">here</a></li>
<li>Description: <code>make check-no-changes</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-23678/runs/1/steps/175/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z] +Copyright 2009 The Go Authors. All rights reserved.
[2021-01-26T10:44:32.715Z] +Use of this source code is governed by a BSD-style
[2021-01-26T10:44:32.715Z] +license that can be found in the LICENSE file.
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z] +Redistribution and use in source and binary forms, with or without
[2021-01-26T10:44:32.715Z] +modification, are permitted provided that the following conditions are
[2021-01-26T10:44:32.715Z] +met:
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z] +   * Redistributions of source code must retain the above copyright
[2021-01-26T10:44:32.715Z] +notice, this list of conditions and the following disclaimer.
[2021-01-26T10:44:32.715Z] +   * Redistributions in binary form must reproduce the above
[2021-01-26T10:44:32.715Z] +copyright notice, this list of conditions and the following disclaimer
[2021-01-26T10:44:32.715Z] +in the documentation and/or other materials provided with the
[2021-01-26T10:44:32.715Z] +distribution.
[2021-01-26T10:44:32.715Z] +   * Neither the name of Google Inc. nor the names of its
[2021-01-26T10:44:32.715Z] +contributors may be used to endorse or promote products derived from
[2021-01-26T10:44:32.715Z] +this software without specific prior written permission.
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z] +THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
[2021-01-26T10:44:32.715Z] +"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
[2021-01-26T10:44:32.715Z] +LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
[2021-01-26T10:44:32.715Z] +A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
[2021-01-26T10:44:32.715Z] +OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
[2021-01-26T10:44:32.715Z] +SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
[2021-01-26T10:44:32.715Z] +LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
[2021-01-26T10:44:32.715Z] +DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
[2021-01-26T10:44:32.715Z] +THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
[2021-01-26T10:44:32.715Z] +(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
[2021-01-26T10:44:32.715Z] +OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z] +
[2021-01-26T10:44:32.715Z]  --------------------------------------------------------------------------------
[2021-01-26T10:44:32.715Z]  Dependency : github.com/cloudfoundry-community/go-cfclient
[2021-01-26T10:44:32.715Z]  Version: v0.0.0-20190808214049-35bcce23fc5f
[2021-01-26T10:44:32.715Z] diff --git a/go.sum b/go.sum
[2021-01-26T10:44:32.715Z] index 7e96acfee9..01efacf251 100644
[2021-01-26T10:44:32.715Z] --- a/go.sum
[2021-01-26T10:44:32.715Z] +++ b/go.sum
[2021-01-26T10:44:32.715Z] @@ -155,9 +155,6 @@ github.com/cespare/xxhash/v2 v2.1.1/go.mod h1:VGX0DQ3Q6kWi7AoAeZDth3/j3BFtOZR5XL
[2021-01-26T10:44:32.715Z]  github.com/chzyer/logex v1.1.10/go.mod h1:+Ywpsq7O8HXn0nuIou7OrIPyXbp3wmkHB+jjWRnGsAI=
[2021-01-26T10:44:32.716Z]  github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e/go.mod h1:nSuG5e5PlCu98SY8svDHJxuZscDgtXS6KTTbou5AhLI=
[2021-01-26T10:44:32.716Z]  github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1/go.mod h1:Q3SI9o4m/ZMnBNeIyt5eFwwo7qiLfzFZmjNmxjkiQlU=
[2021-01-26T10:44:32.716Z] -github.com/clbanning/mxj v1.8.4 h1:HuhwZtbyvyOw+3Z1AowPkU87JkJUSv751ELWaiTpj8I=
[2021-01-26T10:44:32.716Z] -github.com/clbanning/mxj/v2 v2.3.2 h1:DSkU65zfrBHtrggxd54X9pK1z/Lw2OwSW5D8p+x1toE=
[2021-01-26T10:44:32.716Z] -github.com/clbanning/mxj/v2 v2.3.2/go.mod h1:hNiWqW14h+kc+MdF9C6/YoRfjEJoR3ou6tn/Qo+ve2s=
[2021-01-26T10:44:32.716Z]  github.com/clbanning/mxj/v2 v2.4.0 h1:DdG6ym6l1bbNr/wDFhNSdt7kGNcBGoa6q0SWXhkFibc=
[2021-01-26T10:44:32.716Z]  github.com/clbanning/mxj/v2 v2.4.0/go.mod h1:hNiWqW14h+kc+MdF9C6/YoRfjEJoR3ou6tn/Qo+ve2s=
[2021-01-26T10:44:32.716Z]  github.com/client9/misspell v0.3.4/go.mod h1:qj6jICC3Q7zFZvVWo7KLAzC3yx5G7kyvSDkc90ppPyw=
[2021-01-26T10:44:33.652Z] NOTICE.txt: needs update
[2021-01-26T10:44:33.652Z] go.sum: needs update
[2021-01-26T10:44:33.652Z] Makefile:116: recipe for target 'check-no-changes' failed
[2021-01-26T10:44:33.652Z] make: *** [check-no-changes] Error 1
[2021-01-26T10:44:34.038Z] Client: Docker Engine - Community
[2021-01-26T10:44:34.038Z]  Version:           20.10.2
[2021-01-26T10:44:34.038Z]  API version:       1.41
[2021-01-26T10:44:34.038Z]  Go version:        go1.13.15
[2021-01-26T10:44:34.038Z]  Git commit:        2291f61
[2021-01-26T10:44:34.038Z]  Built:             Mon Dec 28 16:17:32 2020
[2021-01-26T10:44:34.038Z]  OS/Arch:           linux/amd64
[2021-01-26T10:44:34.038Z]  Context:           default
[2021-01-26T10:44:34.038Z]  Experimental:      true
[2021-01-26T10:44:34.038Z] 
[2021-01-26T10:44:34.038Z] Server: Docker Engine - Community
[2021-01-26T10:44:34.038Z]  Engine:
[2021-01-26T10:44:34.038Z]   Version:          20.10.2
[2021-01-26T10:44:34.038Z]   API version:      1.41 (minimum version 1.12)
[2021-01-26T10:44:34.038Z]   Go version:       go1.13.15
[2021-01-26T10:44:34.038Z]   Git commit:       8891c58
[2021-01-26T10:44:34.038Z]   Built:            Mon Dec 28 16:15:09 2020
[2021-01-26T10:44:34.038Z]   OS/Arch:          linux/amd64
[2021-01-26T10:44:34.038Z]   Experimental:     false
[2021-01-26T10:44:34.038Z]  containerd:
[2021-01-26T10:44:34.038Z]   Version:          1.4.3
[2021-01-26T10:44:34.038Z]   GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
[2021-01-26T10:44:34.038Z]  runc:
[2021-01-26T10:44:34.038Z]   Version:          1.0.0-rc92
[2021-01-26T10:44:34.038Z]   GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
[2021-01-26T10:44:34.038Z]  docker-init:
[2021-01-26T10:44:34.038Z]   Version:          0.19.0
[2021-01-26T10:44:34.038Z]   GitCommit:        de40ad0
[2021-01-26T10:44:34.038Z] Unable to find image 'alpine:3.4' locally
[2021-01-26T10:44:34.605Z] 3.4: Pulling from library/alpine
[2021-01-26T10:44:34.881Z] c1e54eec4b57: Pulling fs layer
[2021-01-26T10:44:35.143Z] c1e54eec4b57: Download complete
[2021-01-26T10:44:35.402Z] c1e54eec4b57: Pull complete
[2021-01-26T10:44:35.402Z] Digest: sha256:b733d4a32c4da6a00a84df2ca32791bb03df95400243648d8c539e7b4cce329c
[2021-01-26T10:44:35.402Z] Status: Downloaded newer image for alpine:3.4
[2021-01-26T10:44:40.185Z] Stage "Build&Test" skipped due to earlier failure(s)
[2021-01-26T10:44:40.257Z] Stage "Packaging" skipped due to earlier failure(s)
[2021-01-26T10:44:40.350Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-23678/src/github.com/elastic/beats
[2021-01-26T10:44:41.265Z] Running on worker-1244230 in /var/lib/jenkins/workspace/Beats_beats_PR-23678
[2021-01-26T10:44:41.384Z] [INFO] getVaultSecret: Getting secrets
[2021-01-26T10:44:41.485Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-01-26T10:44:43.603Z] + chmod 755 generate-build-data.sh
[2021-01-26T10:44:43.603Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23678/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23678/runs/1 FAILURE 805301
[2021-01-26T10:44:43.603Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23678/runs/1/steps/?limit=10000 -o steps-info.json
[2021-01-26T10:44:44.300Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-23678/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-01-26T10:44:44.998Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-01-26T10:44:46.443Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

jsoriano commented on libbeat/processors/xmldecode/xmldecode_test.go:
```diff
@@ -0,0 +1,255 @@
+			},
+			error:        true,
+			errorMessage: "error trying to decode XML field xml.Decoder.Token() - XML syntax error on line 7: element <book> closed by </ook>",
+		},
```
Add tests for some of the other options? And a case where the field doesn't exist?

jsoriano reviewed (commented):
Thanks! This can be a nice addition. I have added some questions and suggestions.

P1llus commented on libbeat/common/xml.go:
```diff
@@ -0,0 +1,25 @@
+func UnmarshalXML(body []byte, prepend bool, toLower bool) (obj map[string]interface{}, err error) {
+	var xmlobj mxj.Map
+	// Disables attribute prefixes and forces all lines to lowercase to meet ECS standards
+	mxj.PrependAttrWithHyphen(prepend)
```
Yeah this one was currently hardcoded because I couldn't find any reason why we wanted to have all key names starting with a hyphen, it would be easy to change later down the line though.

jsoriano replied:
Ok, let's leave it as is by now and we'll see if there are cases that need some other handling :+1:

jsoriano replied:
Btw, my thoughts were not about leaving the fields with an hyphen in the final event, but about maybe putting the attributes under a different namespace, so they can be distinguished from children objects.

P1llus commented on libbeat/processors/xmldecode/xmldecode.go:
```diff
@@ -0,0 +1,162 @@
+			event.Meta[events.FieldMetaID] = id
+		}
+	}
+	// If jsontransform has not set an error, and it happened elsewhere, add it to the error
```
It's actually referring to the jsontransform function on line 119, its reusing the already built-in function used by decode_json_fields to overwrite, expand and append errors to keys.

jsoriano replied:
Oh ok, I see, all good then :+1:

P1llus commented on libbeat/processors/xmldecode/xmldecode.go:
```diff
@@ -0,0 +1,162 @@
+)
+
+func init() {
+	processors.RegisterPlugin("xmldecode",
```
No problem, I was actually mirroring the `urldecode` processor, so I followed that naming convention, but I don't mind user the other one.

jsoriano replied:
Ok, I think this is more in the line of `decode_json_fields` than `urldecode` :slightly_smiling_face:

P1llus commented on libbeat/processors/xmldecode/xmldecode.go:
```diff
@@ -0,0 +1,162 @@
+			continue
+		}
+
+		xmloutput, err := x.decodeField(field, data)
```
This part of the code is exactly how` decode_json_fields` performs the action though @jsoriano ? Is it not so that it does not completely fail if one of multiple fields does not exist?

jsoriano replied:
Before unmarshalling, `decode_json_fields` checks if the data is actually a string: https://github.com/elastic/beats/blob/3e47a781bb0e11a04caed9ab87cad5ad6ea80350/libbeat/processors/actions/decode_json_fields.go#L107-L121

This effectively prevents trying to decode `null` if `err == common.ErrKeyNotFound`, and is in general safer in case `data` is of any other non-string type.

With this you would also have in `text` a string, so you wouldn't need the `Sprintf` in https://github.com/elastic/beats/pull/23678/files#diff-90271196007e26ceaf04e257f546933b4a3c553145831bfd695559bd4bbc1ef5R145.

jsoriano commented on libbeat/processors/xmldecode/xmldecode_test.go:
```diff
@@ -0,0 +1,255 @@
+			Output: common.MapStr{
+				"message": map[string]interface{}{
+					"catalog": map[string]interface{}{
+						"book": []interface{}{
```
Oh, one important thing I have just recalled. This case of arrays of objects won't be actually work as expected, this is not supported in Elasticsearch, see the warning in https://www.elastic.co/guide/en/elasticsearch/reference/7.10/array.html

You can read more about this in the JSON case on this comment: https://github.com/elastic/beats/pull/21178#issuecomment-705464067

The solution would probably be to generate multiple events, but this cannot be done with the current processors framework. It can be done at the processor level, for example the `httpjson` input has options to support these cases generating multiple events from single objects or arrays.

I mean, not much change here, but just take into account that a processor shouldn't be used to decode arrays of objects. Only thing would be to modify the test cases to include examples with single objects, that is also the main use case for `decode_json_fields`.

P1llus replied:
Array of objects would still be possible to ingest and visualize in some manners, I don't think we should error on it or not allow it at all.

The processor will just decode what is sent to it, and indeed the splitting will happen on the input side.
An array of objects still has certain usecases, and they might also want to include it in the pipeline solely so that you can use ingest nodes to iterate over them and for example append them to a field.

The test cases already covers both single and multiple object usecases, would there be any reason not to @jsoriano ?

andrewstucki replied:
Just to clarify, arrays of objects are perfectly fine and can even be queried independently if you're working with `nested` or `flattened` types -- I can very much imagine that as a use-case for this processor if you're wanting to model an event with nested information. So I don't think we should necessarily give any guidance that you shouldn't use this processor for arrays of data, it all just depends on what you're modeling.

- If you want to model an array as multiple events, use the input
- If you want to model `nested` or `flattened` fields, this processor would work great for that

jsoriano replied:
Agree, my comment was more for ourselves, to be clear that arrays of objects is a use case that usually don't work as expected, without a lot of care.
Mappings with `nested` or `flattened` fields effectively would work, but this is not something that we are frequently using in beats, and is something not so straightforward to use in general.

jsoriano replied:
I didn't know there have been some improvements in kibana during 7.x for nested fields https://github.com/elastic/kibana/issues/1084#issuecomment-585178079, perhaps we could use them more, yes.

jsoriano commented on libbeat/processors/decode_xml_fields/decode_xml_fields.go:
```diff
@@ -0,0 +1,164 @@
+	processors.RegisterPlugin("decode_xml_fields",
+		checks.ConfigChecked(NewDecodeXMLFields,
+			checks.RequireFields("fields"),
+			checks.AllowedFields("fields", "overwrite_keys", "add_error_key", "expand_keys", "target", "document_id")))
```
`expand_keys` has been removed.

jsoriano reviewed (approved):
One minor thing, for the rest it LGTM.

elasticmachine commented:
Pinging @elastic/agent (Team:Agent)

P1llus commented:
Started off another build, as some CI's are failing, but did not seem to be because of my changes.

P1llus commented:
jenkins run tests please

P1llus commented:
jenkins run tests please

P1llus commented:
jenkins run tests please

P1llus commented:
jenkins run tests

jsoriano commented:
/package

P1llus commented:
@jsoriano do you have any idea how we can make these builds pass? Iwe rerun the builds about 6 times this week, last week the kibana snapshot made the builds fail and this week I am still a bit unsure.

andrewkroh commented on libbeat/processors/decode_xml_fields/docs/decode_xml_fields.asciidoc:
```diff
@@ -0,0 +1,74 @@
+[[decode_xml_fields]]
```
I think it would be nice to have small example showing the input/output of this processor.

andrewkroh commented on libbeat/processors/decode_xml_fields/decode_xml_fields.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// NewDecodeXMLFields construct a new decode_xml_fields processor.
+func NewDecodeXMLFields(c *common.Config) (processors.Processor, error) {
```
Add a `cfgwarn.Experimental()` statement since this is experimental.

andrewkroh reviewed (changes requested):
Looks like a nice addition. I'd like to take a look at it again after you address the comments.

urso commented on libbeat/common/xml.go:
```diff
@@ -0,0 +1,42 @@
+package common
+
+import (
+	"github.com/clbanning/mxj/v2"
```
Why not [encoding/xml](https://pkg.go.dev/encoding/xml) ?

P1llus replied:
`encoding/xml ` does not really support the usecase unfortunately, the library I currently use is for the unmarshalling into a unknown struct/interface{}.

If this is something they have added recently I don't mind changing it if needed.

urso commented on libbeat/common/xml.go:
```diff
@@ -0,0 +1,42 @@
+
+// UnmarshalXML takes a slice of bytes, and returns a map[string]interface{}.
+// If the slice is not valid XML, it will return an error.
+func UnmarshalXML(body []byte, prepend bool, toLower bool) (obj map[string]interface{}, err error) {
```
Do we need this function somewhere else in the future, or is it specific to the processor being implemented here?

`libbeat/common` is our sin bin for almost everything. Let's try to not grow it if possible. If we need decoding helpers or similar we should introduce appropriate packages.

P1llus replied:
The general idea is to be able to reuse it (if we want to) on other inputs, I added a small section on it in the PR post:

`"It adds a small XML Unmarshal helper function to the common folder in Libbeat, the reason this is added is so that not only processors can utilize it if/when needed, but also inputs. For example I plan to reuse this in http_endpoint to add XML support there as well. Other beats/inputs that would benefit from this is Winlogbeat."`

P1llus replied:
After a discussion with Andrew I will make this a subpackage of the processor instead, which will make it cleaner and more similar to `decode_cef`

urso commented on libbeat/processors/decode_xml_fields/decode_xml_fields.go:
```diff
@@ -0,0 +1,164 @@
+)
+
+type decodeXMLFieldsConfig struct {
+	Fields        []string `config:"fields"`
```
+1

urso commented on libbeat/processors/decode_xml_fields/decode_xml_fields.go:
```diff
@@ -0,0 +1,164 @@
+	Fields        []string `config:"fields"`
+	OverwriteKeys bool     `config:"overwrite_keys"`
+	AddErrorKey   bool     `config:"add_error_key"`
+	Target        *string  `config:"target"`
```
why pointer? What if `Target` is the empty string?

P1llus replied:
This is mostly because I was following `decode_json_fields`, which implements it similarly, and also supports empty string as target field.

I also didn't really see a usage for it, but kept it for the sake of being as similar as possible to the other processor.

P1llus replied:
@urso After getting some background info from Andrew, I see now why it was a pointer in the older decode_json_fields processor, I will change this to not be a pointer anymore, it was mostly to detect the empty string if nothing was set, but better to set it as part of the default config.

P1llus replied:
I ended up having to revert this, as it was not really feasible to differentiate between nil and an empty string without using it as a pointer.

P1llus commented:
Adding in the last changes, though still a few more left before a second review is ready.

P1llus commented:
@andrewkroh 
Added in the changes you requested, I also moved Target from single string to pointer again, as there was no easily or feasible way to determine if the value was null or empty string (for adding messages to root).

urso commented on CHANGELOG.next.asciidoc:
```diff
@@ -602,8 +602,10 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
+- Added new decode_xml_fields processor to libbeat, that is available to all beat types. {pull}23678[23678]
 - Add deployment name in pod's meta. {pull}23610[23610]
 - Add `selector` information in kubernetes services' metadata. {pull}23730[23730]
+- Add XML processor to libbeat. {pull}23678[23678]
```
I don't think we need 2 changelog entries.

P1llus commented:
jenkins run tests please

P1llus commented:
jenkins run tests please

andrewkroh commented on libbeat/processors/decode_xml/config.go:
```diff
@@ -0,0 +1,39 @@
+	return decodeXMLConfig{
+		Field:         "message",
+		IgnoreMissing: false,
+		Target:        new(string),
```
This conflicts with the default described in the documentation. Leave this `nil` to make replacing `Field` be the default

andrewkroh commented on libbeat/processors/decode_xml/docs/decode_xml.asciidoc:
```diff
@@ -0,0 +1,110 @@
+exist in the event are overwritten by keys from the decoded XML object. The
+default value is false.
+
+`add_error_key`:: (Optional) If it set to true, in case of error while decoding XML objects
```
No longer an option.

andrewkroh commented on libbeat/processors/decode_xml/docs/decode_xml.asciidoc:
```diff
@@ -0,0 +1,110 @@
+-------------------------------------------------------------------------------
+
+
+In the examples above:
```
How about "The supported configuration options are:". This way it's clear that we're provided reference info.

andrewkroh commented on libbeat/processors/decode_xml/docs/decode_xml.asciidoc:
```diff
@@ -0,0 +1,110 @@
+`document_id`:: (Optional) XML key to use as the document id. If configured,
+the field will be removed from the original XML document and stored in
+`@metadata._id`
+
```
Some configuration options are missing: `field`, `ignore_failure`, `ignore_missing`.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+				<secondcategory>
+					<paper id="bk102">
+						<test2>Ralls, Kim</test2>
+						<description>A former architect battles corporate zombies, 
```
Please remove all lines with trailing spaces. A lot people's editors do this automatically so these will cause lots of problems.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+				catalog>`,
+			},
+			Output: common.MapStr{
+				"message": (map[string]interface{})(nil),
```
Where did `message` go? It failed and it deleted the source data. That seems bad.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+				"message": (map[string]interface{})(nil),
+			},
+			error:        true,
+			errorMessage: "failed to decode fields in decode_xml processor: error decoding XML field: xml.Decoder.Token() - XML syntax error on line 7: element <book> closed by </ook>",
```
`ignore_failure == true` for this test so why was an error returned.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+				"error":   "key not found; field value is not a string",
+			},
+			error:        true,
+			errorMessage: "key not found; field value is not a string",
```
There are two errors here. I would expect only the `key not found` error. After that it should stop processing. Additionally it would nice to include the name of the missing key for context.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+		},
+	}
+
+	for _, test := range testCases {
```
Nice test coverage here. 👍

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+		"target":      "message",
+	})
+
+	p, err := New(config)
```
Use the `newDecodeXML` constructor.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+
+	p, err := New(config)
+	if err != nil {
+		log.Error("Error initializing decode_xml")
```
Avoid relying on `logp` in tests. If you need to log something from a tests use `t.Log`. But in this case you already have `t.Fatal` so just remove it.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml_test.go:
```diff
@@ -0,0 +1,482 @@
+		test := test
+		t.Run(test.description, func(t *testing.T) {
+			t.Parallel()
+			f := &decodeXML{
```
Use `newDecodeXML` to ensure all initialization is done.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml.go:
```diff
@@ -0,0 +1,156 @@
+			errs = append(errs, fmt.Errorf("failed to decode fields in decode_xml processor: %v", err))
+		}
+
+		target := field
```
This logic could be move to the constructor such that `Target` is set once based on the behavior you define.

andrewkroh commented on libbeat/processors/decode_xml/decode_xml.go:
```diff
@@ -0,0 +1,156 @@
+		if x.config.IgnoreMissing && err == common.ErrKeyNotFound {
+			return event, nil
+		}
+		errs = append(errs, err)
```
The error handling is still kind of complicated here. I think it's because you want to handle the `IgnoreFailure` logic in one place. If you move the processing into a separate function that always returns an error on failure it will be simple to handle the `IgnoreFailure` logic in one place.

andrewkroh reviewed (changes requested):
The problem I found with the third-party library makes me wonder if we ought to fix that first before adding this processor.

I'm attaching a patch to demonstrate some of the changes (not all) that I made in comments. You can use `git apply <file>` to try it.

[0001-Simplify-error-handling-and-fix-race.patch.txt](https://github.com/elastic/beats/files/5969096/0001-Simplify-error-handling-and-fix-race.patch.txt)

elasticmachine commented:
Pinging @elastic/ingest-management (Team:Ingest Management)

P1llus commented:
jenkins run tests please

kuisathaverat commented:
lint stage failed

```
[2021-02-15T08:08:06.836Z] + make -C libbeat check

[2021-02-15T08:08:06.836Z] make: Entering directory '/var/lib/jenkins/workspace/Beats_beats_PR-23678/src/github.com/elastic/beats/libbeat'

[2021-02-15T08:08:13.413Z] >> fmt - go-licenser: Checking for missing headers

[2021-02-15T08:08:15.951Z] >> fmt - go-licenser: Adding missing headers

[2021-02-15T08:08:15.951Z] >> fmt - autopep8: Formatting Python code

[2021-02-15T08:08:15.951Z] >> fmt - goimports: Formatting Go code

[2021-02-15T08:08:42.504Z] >> check: Checking source code for common problems

[2021-02-15T08:09:21.252Z] # github.com/elastic/beats/v7/libbeat/common/encoding/xml

[2021-02-15T08:09:21.252Z] common/encoding/xml/decode_test.go:302:1: ExampleXMLToJSON refers to unknown identifier: XMLToJSON

[2021-02-15T08:09:26.523Z] Error: failed running go vet, please fix the issues reported: running "go vet ./..." failed with exit code 2

[2021-02-15T08:09:26.524Z] scripts/Makefile:152: recipe for target 'check' failed

[2021-02-15T08:09:26.524Z] make: *** [check] Error 1

[2021-02-15T08:09:26.524Z] make: Leaving directory '/var/lib/jenkins/workspace/Beats_beats_PR-23678/src/github.com/elastic/beats/libbeat'
```

P1llus commented:
Thanks @kuisathaverat . Sorry for the build spam, it has had quite some issues with failing test runs the whole week, mostly unrelated however this one is related for sure.

P1llus commented:
It seems to be related to an earlier commit, as there is no reference to those identifiers anymore. Locally the test also runs fine, but will look to see if existed initially and was removed in a follow-up commit:

```
> make -C libbeat check
make: Entering directory '/home/marius/go/src/github.com/elastic/beats/libbeat'
Installing mage v1.11.0.
/home/marius/.magefile cleaned
>> fmt - go-licenser: Checking for missing headers
>> fmt - go-licenser: Adding missing headers
>> fmt - goimports: Formatting Go code
>> fmt - autopep8: Formatting Python code
>> check: Checking source code for common problems
make: Leaving directory '/home/marius/go/src/github.com/elastic/beats/libbeat'
```

P1llus commented:
The naming convention of the Example test does not match, will push a update.
